webpush/push.go

```diff
@@ -32,24 +32,36 @@ const (
 // (notably Google Cloud Messaging, used by Chrome) then you can add that as the
 // token parameter.
 func NewPushRequest(sub *Subscription, message string, token string) (*http.Request, error) {
-	payload, err := Encrypt(sub, message)
-	if err != nil {
-		return nil, err
-	}
-
 	// If the endpoint is GCM then we temporarily need to rewrite it, as not all
 	// GCM servers support the Web Push protocol. This should go away in the
 	// future.
 	endpoint := strings.Replace(sub.Endpoint, gcmURL, tempGcmURL, 1)
 
-	req, err := http.NewRequest("POST", endpoint, bytes.NewReader(payload.Ciphertext))
-	if err != nil {
-		return nil, err
-	}
+	var req *http.Request
+	var err error
+
+	// If there is no payload then we don't actually need encryption
+	if message != "" {
+		var payload *EncryptionResult
+		payload, err = Encrypt(sub, message)
+		if err != nil {
+			return nil, err
+		}
 
-	req.Header.Add("Encryption", headerField("salt", payload.Salt))
-	req.Header.Add("Crypto-Key", headerField("dh", payload.ServerPublicKey))
-	req.Header.Add("Content-Encoding", "aesgcm")
+		req, err = http.NewRequest("POST", endpoint, bytes.NewReader(payload.Ciphertext))
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Add("Encryption", headerField("salt", payload.Salt))
+		req.Header.Add("Crypto-Key", headerField("dh", payload.ServerPublicKey))
+		req.Header.Add("Content-Encoding", "aesgcm")
+	} else {
```

Comment:
Just a thought: wouldn't it be more readable if the `if` conditions were reversed?

```go
req, err := http.NewRequest("POST", endpoint, nil)
if err != nil {
  return nil, err
}
if token != "" {
  req.Header.Add("Authorization", "key="+token)
}
if message == "" {
  return req, nil
}

// encrypted payload
payload, err := Encrypt(sub, message)
if err != nil {
  return nil, err
}
req.Body = ioutil.NopCloser(bytes.NewReader(payload.Ciphertext))
req.ContentLength = len(payload.Ciphertext)
req.Header.Add("Encryption", headerField("salt", payload.Salt))
req.Header.Add("Crypto-Key", headerField("dh", payload.ServerPublicKey))
req.Header.Add("Content-Encoding", "aesgcm")
return req, nil
```